Jetty 12 - #8723: Provide a thread-safe way to modify HttpClient proxies

cowwoc opened the pull request:
…at runtime.

Unlike the v10, v11 implementations I opted to use `BlockingQueue` instead of `CopyOnWriteList`. This should improve performance.

cowwoc committed:
Fixes #8723: Provide a thread-safe way to modify HttpClient proxies at runtime.

cowwoc committed:
Fixing checkstyle errors.

cowwoc committed:
Fixed unit tests.

cowwoc committed:
Fixed unit tests and documentation.

cowwoc committed:
Fixed unit tests.

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/ProxyConfiguration.java:
```diff
@@ -30,23 +32,42 @@
 public class ProxyConfiguration
 {
-    private final List<Proxy> proxies = new ArrayList<>();
+    private final BlockingQueue<Proxy> proxies = new LinkedBlockingQueue<>();
```
```suggestion
    private final List<Proxy> proxies = new BlockingArrayQueue<>();
```

sbordet commented on jetty-ee9/jetty-ee9-proxy/src/test/java/org/eclipse/jetty/ee9/proxy/ProxyServletTest.java:
```diff
@@ -123,6 +126,7 @@ public static Stream<Arguments> impls()
     }
 
     private HttpClient client;
+    private final List<Proxy> clientProxies = new ArrayList<>();
```
This field can be just a `Proxy proxy`.

sbordet commented on jetty-ee9/jetty-ee9-proxy/src/test/java/org/eclipse/jetty/ee9/proxy/ProxyServletTest.java:
```diff
@@ -197,16 +201,18 @@ private void startClient() throws Exception
     {
-        client = prepareClient(consumer);
+        clientProxies.add(new HttpProxy("localhost", proxyConnector.getLocalPort()));
+        client = prepareClient(consumer, clientProxies);
```
Just initialize the `proxy` field, no need to pass it around, as `prepareClient()` will have access to it.

gregw committed:
Implement ErrorPageErrorProcessor to resolve #8726 (#8736)

sbordet committed:
Code cleanups.

Removed obsolete oej.server.HttpTransport.
Cleaned up ConnectionMetaData.Wrapper.
Simplified ee10 DefaultServlet.ServletCoreRequest.
Other minor cleanups.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

cowwoc committed:
Update in response to PR comments.

sbordet commented:
@cowwoc thanks!
Resolved, squashed and merged with 67c36207c211a163b07187bbfc1e8c01696a8030.
